### PR TITLE
Fix roundtrip conversions between v1alpha2 and v1alpha3

### LINKED ISSUE #2059: v1alpha2/v1alpha3 skew - Bastion does not have Public IP in AWSCluster status
/kind bug

**What steps did you take and what happened:**
1. Applied v1alpha3 webhooks and CRDs on a management cluster
2. Created a cluster using v1alpha2 controllers CAPA 0.4.10 and CAPI 0.2.11 and CABPK 0.1.7
3. Cluster creation was successful.
4. Bastion was created in AWS, with Public IP.

**What happened:**
The AWSCluster status did not have the Public IP of the bastion. The instance state was pending.

**What did you expect to happen:**
Expected the Public IP of the bastion in the AWSCluster status. 
Expected the instance state to be running

**Anything else you would like to add:**
CURLed to v1alpha2 endpoint of AWSCluster object, and patched --data '{"status":{"bastion":{"instanceState":"running"}}}',  the data was not changed in AWSCluster,
CURLed the same data to v1alpha3 endpoint , and the instance state changed to running

**Environment:**

- Cluster-api-provider-aws version:  CAPA 0.4.10 and CAPI 0.2.11 and CABPK 0.1.7
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by sedefsavas**
**What this PR does / why we need it**:

This PR prevents overriding ControlPlaneLoadBalancer and Bastion changes.
It also avoids creating `CloudInit` during the conversion if it is empty, because if we create an empty `CloudInit` during conversion,
`out.EnableSecureSecretsManager = !in.InsecureSkipSecretsManager` sets a field that should not be there.
``

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2059

### CODE REVIEW

**Comment by sedefsavas** [comment]
cc @ncdc

**Comment by randomvariable** [comment]
Are there some tests we could add that would have picked up on the issues?

**Comment by ncdc** [comment]
@randomvariable yes, we need to make the change in CAPI though

**Comment by sedefsavas** [comment]
Added a reverse fuzz test: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2074

I could add individual tests as well for the 3 fields I changed here.

**Comment by randomvariable** [comment]
If an upstream CAPI change is going to deal with it, that's fine.

/lgtm

I'll leave @ncdc to do the approve given the greater context you have.

File: api/v1alpha2/awsmachine_conversion.go [hunk 1]
```diff
@@ -70,9 +73,23 @@ func restoreAWSMachineSpec(restored, dst *infrav1alpha3.AWSMachineSpec) {
...
+		}
+	}
 
 	dst.Tenancy = restored.Tenancy
-	dst.CloudInit.SecureSecretsBackend = restored.CloudInit.SecureSecretsBackend
+
+	if restored.CloudInit.SecureSecretsBackend != "" {
+		if src.CloudInit != nil {
+			dst.CloudInit.SecureSecretsBackend = restored.CloudInit.SecureSecretsBackend
+		}
```
Comment (randomvariable) [inline api/v1alpha2/awsmachine_conversion.go:91]: Back when I added the secure secrets stuff, I had originally also added 
image before removing it thinking it was excessive.

Do we in fact need to conditionally restore the SecretARN here too?

Comment (ncdc) [reply api/v1alpha2/awsmachine_conversion.go:91]: SecretARN turned into SecretPrefix + SecretCount.

Comment (ncdc) [reply api/v1alpha2/awsmachine_conversion.go:91]: And SecretPrefix + SecretCount exist in both a2 and a3

**Comment by randomvariable** [comment]
/retitle Fix roundtrip conversions between v1alpha2 and v1alpha3
/approve

**Pull request merged**